src/arch/generic/semihosting.hh

```diff
@@ -0,0 +1,551 @@
+/*
+ * Copyright (c) 2018, 2019 ARM Limited
+ * All rights reserved
+ *
+ * The license below extends only to copyright in the software and shall
+ * not be construed as granting a license to any other intellectual
+ * property including but not limited to intellectual property relating
+ * to a hardware implementation of the functionality of the software
+ * licensed hereunder.  You may use the software subject to the license
+ * terms below provided that you ensure that this notice is replicated
+ * unmodified and in its entirety in all distributions of the software,
+ * modified or unmodified, in source code or in binary form.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met: redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer;
+ * redistributions in binary form must reproduce the above copyright
+ * notice, this list of conditions and the following disclaimer in the
+ * documentation and/or other materials provided with the distribution;
+ * neither the name of the copyright holders nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef __ARCH_GENERIC_SEMIHOSTING_HH__
+#define __ARCH_GENERIC_SEMIHOSTING_HH__
+
+#include <cstdio>
+#include <functional>
+#include <map>
+#include <memory>
+#include <utility>
+#include <vector>
+
+#include "base/time.hh"
+#include "cpu/thread_context.hh"
+#include "mem/port_proxy.hh"
+#include "sim/core.hh"
+#include "sim/guest_abi.hh"
+#include "sim/sim_object.hh"
+
+namespace gem5
+{
+
+struct BaseSemihostingParams;
+class SerialDevice;
+
+/**
+ * Semihosting for AArch32, AArch64, RISCV-32 and RISCV-64
```

Comment:
I believe the intention behind the RISC-V semihosting spec was to be able to reuse as much as possible from existing ARM implementations (both emulators as well as libc support). It's not a final spec, so I believe it's still possible to make some changes to it to remove old quirks (although QEMU has implemented the current version since 6.0 so not sure if there is much that can be changed at this point).

CC: @keith-packard

Reply:
Yup, the goal was to share the same spec and implementations to avoid fragmentation. I just inherited the bits from the actual authors (mostly Palmer and Megan, I believe?) and got them published so I could upstream the qemu support and use that for picolibc testing.